feat(password): add new component (#5837)
Sage/carbon: src/components/password/password.stories.mdx

### With label Align

<Canvas>
  <Story name="with labelAlign" story={stories.WithLabelAlign} />
</Canvas>

### With custom labelWidth and inputWidth

<Canvas>
  <Story name="with custom labelWidth and inputWidth" story={stories.WithCustomLabelWidthAndInputWidth} />
</Canvas>

### With custom maxWidth

<Canvas>
  <Story name="with custom maxWidth" story={stories.WithCustomMaxWidth} />
</Canvas>

### With fieldHelp

<Canvas>
  <Story name="with fieldHelp" story={stories.WithFieldHelp} />
</Canvas>

### With labelHelp

<Canvas>
  <Story name="with labelHelp" story={stories.WithLabelHelp} />
</Canvas>

### With required

<Canvas>
  <Story name="required" story={stories.WithRequired} />
</Canvas>

### Validations

Validation status can be set by passing `error`, `warning` or `info` prop to the component.

Passing a string to these props will display a properly colored border along with a validation icon and tooltip - string value will be displayed as the tooltip message.

Passing a boolean to these props will display only a properly colored border.

For more information check our [Validations](?path=/docs/documentation-validations--string-validation "Validations") documentation page.

> 

Reviewer: You'll need to add examples here of how `Password` should look with validations. Old and new validations like we have with `Decimal`.
https://carbon.sage.com/?path=/docs/decimal-input--validations-string-component#validations

## Props

### Password

> Due to the `Textbox` component being used internally by the `Password` component all the props of `Textbox` can be applied to `Password` as well.

Author: In my opinion, we should be surfacing `type` because consumers will want to enter their own mechanisms to change the `type` back to `password` on submission etc.

import { Meta, Story, Canvas } from "@storybook/addon-docs";
import LinkTo from "@storybook/addon-links/react";
import StyledSystemProps from "../../../.storybook/utils/styled-system-props";
import Password from ".";
import * as stories from "./password.stories.tsx";

<Meta title="Password" parameters={{ info: { disable: true } }} />

# Password

Provide a way for the user to securely enter a password.

## Contents

- [Quick Start](#quick-start)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import Password from "carbon-react/lib/components/password";
```

## Examples

### Default

<Canvas>
  <Story name="default" story={stories.Default} />
</Canvas>

> ### With other input components used to change type

Reviewer 2: ```suggestion
### Controlling visibility of password
```
What do you think about this wording tweak? This might make it clearer about what this pattern is about at a glance from a design perspective

Please ensure that the input type is set to `password` on form submission, this will ensure browsers do not remember
the inputs content. We have set the `autocomplete` attribute to `off` on the component, but this is not widely 
supported across different browsers and can have unpredictable results.

Also, we ** strongly ** recommend the use of Carbon components: `Switch`, `Checkbox` and `Button` to change the input type of the `Password` component,
we have provided an example below of how this is possible. Please ensure `aria-controls` are used with the `id` prop 
and chosen string to ensure there is a clear indication that the chosen input component is responsible for changing type.

> ### Controlling visibility of password with Carbon component

Reviewer 2: ```suggestion
```

<Canvas>

>   <Story name="with switch" story={stories.ChangingType} />

Reviewer 2: Similar to the previous comment, what do you think?

```suggestion
  <Story name="control password visibility" story={stories.ChangingType} />
```

import Password from "carbon-react/lib/components/password";
```

## Examples

### Default

<Canvas>
  <Story name="default" story={stories.Default} />
</Canvas>

### With other input components used to change type

Please ensure that the input type is set to `password` on form submission, this will ensure browsers do not remember
the inputs content. We have set the `autocomplete` attribute to `off` on the component, but this is not widely 
supported across different browsers and can have unpredictable results.

Also, we ** strongly ** recommend the use of Carbon components: `Switch`, `Checkbox` and `Button` to change the input type of the `Password` component,
we have provided an example below of how this is possible. Please ensure `aria-controls` are used with the `id` prop 
and chosen string to ensure there is a clear indication that the chosen input component is responsible for changing type.

### Changing type with Carbon component

<Canvas>
  <Story name="with switch" story={stories.ChangingType} />
</Canvas>

### Overriding aria-live message translations

Various translations can be applied to the `aria-live` region which contains both a shown and hidden message

> variation. Both of these messages will update assistive technology such as screenreaders When 

Reviewer: ```suggestion
variation. Both of these messages will update assistive technology such as screen readers when 
```

</Canvas>

### Margins

<Canvas>
  <Story name="with margins" story={stories.Margins} />
</Canvas>

### Disabled

<Canvas>
  <Story name="disabled" story={stories.Disabled} />
</Canvas>

### ReadOnly

<Canvas>
  <Story name="readOnly" story={stories.ReadOnly} />
</Canvas>

### AutoFocus

<Canvas>
  <Story name="with autoFocus" story={stories.AutoFocus} />
</Canvas>

### With labelInline

<Canvas>
  <Story name="with labelInline" story={stories.WithLabelInline} />
</Canvas>

### With label Align

<Canvas>
  <Story name="with labelAlign" story={stories.WithLabelAlign} />
</Canvas>

### With custom labelWidth and inputWidth

<Canvas>
  <Story name="with custom labelWidth and inputWidth" story={stories.WithCustomLabelWidthAndInputWidth} />
</Canvas>

### With custom maxWidth

<Canvas>
  <Story name="with custom maxWidth" story={stories.WithCustomMaxWidth} />
</Canvas>

### With fieldHelp

<Canvas>
  <Story name="with fieldHelp" story={stories.WithFieldHelp} />
</Canvas>

### With labelHelp

<Canvas>
  <Story name="with labelHelp" story={stories.WithLabelHelp} />
</Canvas>

### With required

<Canvas>
  <Story name="required" story={stories.WithRequired} />
</Canvas>

### Validations

> Validation status can be set by passing `error`, `warning` or `info` prop to the component

Reviewer: ```suggestion
Validation status can be set by passing `error`, `warning` or `info` prop to the component.
```

Passing a string to these props will display a properly colored border along with a validation icon and tooltip - string value will be displayed as the tooltip message.

Passing a boolean to these props will display only a properly colored border.

> For more information check our [Validations](?path=/docs/documentation-validations--string-validation "Validations") documentation page

Reviewer: ```suggestion
For more information check our [Validations](?path=/docs/documentation-validations--string-validation "Validations") documentation page.
```

## Props

### Password

Due to the `Textbox` component being used internally by the `Password` component all the props of `Textbox` can be applied to `Password` as well.

> <StyledSystemProps of={Textbox} noHeader margin />

Reviewer 3: `type` as a prop is not shown in `Props`

import { Meta, Story, Canvas } from "@storybook/addon-docs";
import LinkTo from "@storybook/addon-links/react";
import StyledSystemProps from "../../../.storybook/utils/styled-system-props";
import Password from ".";
import * as stories from "./password.stories.tsx";

<Meta title="Password" parameters={{ info: { disable: true } }} />

# Password

Provide a way for the user to securely enter a password.

## Contents

- [Quick Start](#quick-start)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import Password from "carbon-react/lib/components/password";
```

## Examples

### Default

> Please ensure that the `forceObscurity` prop is `true` on form submission, this will ensure browsers do not remember
> the inputs content. We have set the `autocomplete` attribute to `off` on the component also, but this is not widely 
> supported across different browsers and can have unpredictable results.

Reviewer 2: 1. As we want to highlight this as an important security disclaimer, what do you think about making this paragraph bold?

```suggestion
**Warning - Please ensure that the `forceObscurity` prop is `true` on form submission, this will ensure browsers do not remember the inputs content.** 

We have set the `autocomplete` attribute to `off` on the component also, but this is not widely supported across different browsers and can have unpredictable results.
```

</Canvas>

It is possible to use the `tooltipPosition` to override the default placement of tooltips rendered as part of this component.

<Canvas>
  <Story
    name="validations - string - with tooltipPosition overriden - component"
    story={stories.ValidationsAsAStringWithTooltipCustom}
  />
</Canvas>

#### As a string, displayed on label

<Canvas>
  <Story
    name="validations - string - label"
    story={stories.ValidationsAsAStringDisplayedOnLabel}
  />
</Canvas>

#### New designs validation

<Canvas>
  <Story
    name="validations - string - new design"
    story={stories.NewDesignsValidation}
  />
</Canvas>

It is possible to use the `tooltipPosition` to override the default placement of tooltips rendered as part of this component.

<Canvas>
  <Story
    name="validations - string - with tooltipPosition overriden - label"
    story={stories.ValidationsAsAStringWithTooltipDefault}
  />
</Canvas>

#### As a boolean

<Canvas>
  <Story name="validations - boolean" story={stories.ValidationsAsABoolean} />
</Canvas>

## Props

### Password

<StyledSystemProps of={Password} noHeader margin />

**Any other supplied props will be provided to the underlying HTML input element**

> 

Reviewer 2: Apologies for not spotting this previously @tomdavies73, it would be a good idea to document the new translation keys too (`password.ariaLiveShownMessage` and `password.ariaLiveHiddenMessage`).

We have several examples for other components which should help you with how to structure this
